service/worker/sysworkflow/system_workflow.go

```diff
@@ -22,101 +22,341 @@ package sysworkflow
+func ArchivalDeleteHistoryActivity(ctx context.Context, request ArchiveRequest) error {
+	container := ctx.Value(sysWorkerContainerKey).(*SysWorkerContainer)
+	logger := container.Logger.WithFields(bark.Fields{
+		logging.TagArchiveRequestDomainID:          request.DomainID,
+		logging.TagArchiveRequestWorkflowID:        request.WorkflowID,
+		logging.TagArchiveRequestRunID:             request.RunID,
+		logging.TagArchiveRequestEventStoreVersion: request.EventStoreVersion,
+		logging.TagArchiveRequestLastFirstEventID:  request.LastFirstEventID,
+	})
+	metricsClient := container.MetricsClient
+	if request.EventStoreVersion == persistence.EventStoreVersionV2 {
+		err := persistence.DeleteWorkflowExecutionHistoryV2(container.HistoryV2Manager, request.BranchToken, container.Logger)
+		if err == nil {
+			return nil
+		}
+		op := func() error {
+			return persistence.DeleteWorkflowExecutionHistoryV2(container.HistoryV2Manager, request.BranchToken, container.Logger)
+		}
+		for err != nil && common.IsPersistenceTransientError(err) {
+			err = backoff.Retry(op, common.CreatePersistanceRetryPolicy(), common.IsPersistenceTransientError)
+		}
+		logger.WithError(err).Error("failed to delete history from events v2")
+		metricsClient.IncCounter(metrics.ArchivalDeleteHistoryActivityScope, metrics.SysWorkerDeleteHistoryV2NonRetryableFailures)
+		return errDeleteHistoryActivityDeleteFromV2
+	}
+	deleteHistoryReq := &persistence.DeleteWorkflowExecutionHistoryRequest{
+		DomainID: request.DomainID,
+		Execution: shared.WorkflowExecution{
+			WorkflowId: common.StringPtr(request.WorkflowID),
+			RunId:      common.StringPtr(request.RunID),
+		},
+	}
+	err := container.HistoryManager.DeleteWorkflowExecutionHistory(deleteHistoryReq)
+	if err == nil {
+		return nil
+	}
+	op := func() error {
+		return container.HistoryManager.DeleteWorkflowExecutionHistory(deleteHistoryReq)
+	}
+	for err != nil && common.IsPersistenceTransientError(err) {
+		err = backoff.Retry(op, common.CreatePersistanceRetryPolicy(), common.IsPersistenceTransientError)
+	}
+	logger.WithError(err).Error("failed to delete history from events v1")
+	metricsClient.IncCounter(metrics.ArchivalDeleteHistoryActivityScope, metrics.SysWorkerDeleteHistoryV1NonRetryableFailures)
+	return errDeleteHistoryActivityDeleteFromV1
+}
+
+func nextBlobRetryForever(historyBlobItr HistoryBlobIterator) (*HistoryBlob, error) {
+	result, err := historyBlobItr.Next()
+	if err == nil {
+		return result, nil
+	}
+
+	op := func() error {
+		result, err = historyBlobItr.Next()
+		return err
+	}
+	for err != nil && common.IsPersistenceTransientError(err) {
+		err = backoff.Retry(op, common.CreatePersistanceRetryPolicy(), common.IsPersistenceTransientError)
+	}
+	return result, err
+}
+
+func blobExistsRetryForever(blobstoreClient blobstore.Client, bucket string, key blob.Key) (bool, error) {
+	exists, err := blobstoreClient.Exists(context.Background(), bucket, key)
```

Comment:
you want to have some reasonable timeout set on the context instead of passing a block forever context